File: 0518-coin-change-ii/0518-coin-change-ii.java
class Solution {
    public int change(int amount, int[] coins) {
        int[] dp = new int[amount + 1];
        dp[0] = 1;
        for (int i = 0; i < coins.length; i++) {
            for (int x = 1; x <= amount; x++) {
                if (x >= coins[i]) {
                    dp[x] = dp[x] + dp[x - coins[i]];
                }
            }
        }
        return dp[amount];
    }
}

File: 0518-coin-change-ii/0518-coin-change-ii.cpp
class Solution {
public:
    int dp[301][5001];
    int Find(int idx, int amount, vector<int>&coins, int n){
        if(idx >= n ){
            return (amount == 0) ? 1 : 0;
        }
        if(dp[idx][amount]!=-1)return dp[idx][amount];
        if(amount == 0) return 1;
        if(amount < 0) return 0;
        int left = 0 , right = 0;
        if(amount >= coins[idx]){
            left = Find(idx,amount-coins[idx],coins,n);
        }
        right = Find(idx+1,amount,coins,n);
        return dp[idx][amount] = left + right;
    }
    int change(int amount, vector<int>& coins) {
        int n = coins.size();
        memset(dp,-1,sizeof(dp));
        int ans = Find(0,amount,coins,n);

        return ans;
    }
};